apm-sniffer/apm-toolkit-activation/apm-toolkit-log4j-1.x-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/log/log4j/v1/x/log/GRPCLogAppenderInterceptor.java

```diff
@@ -0,0 +1,114 @@
+import org.apache.skywalking.apm.network.logging.v3.LogData;
+import org.apache.skywalking.apm.network.logging.v3.LogDataBody;
+import org.apache.skywalking.apm.network.logging.v3.TextLog;
+import org.apache.skywalking.apm.network.logging.v3.TraceContext;
+import org.apache.skywalking.apm.network.logging.v3.YAMLLog;
+import org.apache.skywalking.apm.toolkit.common.log.GRPCLogConfig;
+import org.apache.skywalking.apm.toolkit.common.log.LogReportServiceClient;
+import org.slf4j.event.LoggingEvent;
+
+public class GRPCLogAppenderInterceptor implements InstanceMethodsAroundInterceptor {
+
+    private LogReportServiceClient client;
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                             MethodInterceptResult result) throws Throwable {
+        if (Objects.isNull(client)) {
+            client = ServiceManager.INSTANCE.findService(LogReportServiceClient.class);
+            if (Objects.isNull(client)) {
+                return;
+            }
+        }
+        LoggingEvent event = (LoggingEvent) allArguments[0];
+        if (event != null) {
+            LogData logData = transform(event);
+            if (Objects.nonNull(logData)) {
+                client.produce(logData);
+            }
+        }
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+                              Object ret) throws Throwable {
+        return ret;
+    }
+
+    @Override
+    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments,
+                                      Class<?>[] argumentsTypes, Throwable t) {
+
+    }
+
+    private LogData transform(LoggingEvent event) {
+        if (Objects.isNull(event)) {
+            return null;
+        }
+        LogData.Builder logBuilder = LogData.newBuilder()
+                .setTimestamp(event.getTimeStamp())
+                .setService(Config.Agent.SERVICE_NAME)
+                .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                .setTraceContext(TraceContext.newBuilder()
+                        .setTraceId(ContextManager.getGlobalTraceId()).build())
+                .addTags(KeyStringValuePair.newBuilder()
+                        .setKey("level").setValue(event.getLevel().toString()).build())
+                .addTags(KeyStringValuePair.newBuilder()
+                        .setKey("logger").setValue(event.getLoggerName()).build())
+                .addTags(KeyStringValuePair.newBuilder()
+                        .setKey("thread").setValue(event.getThreadName()).build());
+
+        LogDataBody.Builder logDataBodyBuilder = LogDataBody.newBuilder();
+        LogDataBody.ContentCase contentType = Optional.ofNullable(LogDataBody.ContentCase.forNumber(
+                GRPCLogConfig.Plugin.GRPCLog.LOG_BODY_TYPE)).orElse(LogDataBody.ContentCase.TEXT);
```

Comment:
Doesn't log framework provide the content format? Right now, it seems based on manual configuration.

Reply:
The log framework is format the log content in console appended or file appended. In the current position, we got the real text rather than the formatted log.

Reply:
In the current position, we get the real text, and do we need to add an encoder in appender and send it with formatting log content to the client ?